### PR TITLE
Add GetAllHeaders method to golang HTTP filter

### PR DESCRIPTION
**Pull request by willemveerman**
Commit Message: Add GetAllHeaders method to golang HTTP filter 
Additional Description: Adds a new method which returns a copy of the underlying `map[string][]string` object which contains all the request headers.
Risk Level: Low
Testing: Unit test
Docs Changes: none required - doesn't change the configuration params of the plugin
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: contrib/golang/common/go/api/type.go [hunk 1]
```diff
@@ -131,6 +131,9 @@ type HeaderMap interface {
 
 	// RangeWithCopy calls f sequentially for each key and value copied from the map.
 	RangeWithCopy(f func(key, value string) bool)
+
+	// GetAllHeaders returns a copy of the underlying header map
```
Comment (doujiang24) [inline contrib/golang/common/go/api/type.go:135]:
```suggestion
	// GetAllHeaders returns all of headers.
```

File: contrib/golang/filters/http/source/go/pkg/http/type.go [hunk 2]
```diff
@@ -155,6 +155,14 @@ func (h *requestOrResponseHeaderMapImpl) RangeWithCopy(f func(key, value string)
 	}
 }
 
+func (h *requestOrResponseHeaderMapImpl) GetAllHeaders() map[string][]string {
+	h.mutex.Lock()
+	h.initHeaders()
+	copied_headers := h.headers
+	h.mutex.Unlock()
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/go/pkg/http/type.go:162]:
```suggestion
	defer h.mutex.Unlock()
```
and move to line 160, for safety.

File: contrib/golang/filters/http/source/go/pkg/http/type.go [hunk 3]
```diff
@@ -318,6 +326,14 @@ func (h *requestOrResponseTrailerMapImpl) RangeWithCopy(f func(key, value string
 	}
 }
 
+func (h *requestOrResponseTrailerMapImpl) GetAllHeaders() map[string][]string {
+	h.mutex.Lock()
+	h.initTrailers()
+	copied_headers := h.headers
+	h.mutex.Unlock()
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/go/pkg/http/type.go:333]: ditto.

File: contrib/golang/filters/http/source/go/pkg/http/type.go [hunk 4]
```diff
@@ -155,6 +155,14 @@ func (h *requestOrResponseHeaderMapImpl) RangeWithCopy(f func(key, value string)
 	}
 }
 
+func (h *requestOrResponseHeaderMapImpl) GetAllHeaders() map[string][]string {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	h.initHeaders()
+	copied_headers := h.headers
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/go/pkg/http/type.go:162]: this won't copy a map, we should deep clone it.

**Review by doujiang24 — COMMENTED** [review]
Thanks, we need some tests though the implementation is simple.

**Comment by phlax** [comment]
@doujiang24 i think this may be waiting for further review

**Comment by willemveerman** [comment]
> @doujiang24 i think this may be waiting for further review

I still need to add the tests

File: contrib/golang/filters/http/source/go/pkg/http/type.go [hunk 5]
```diff
@@ -318,6 +329,17 @@ func (h *requestOrResponseTrailerMapImpl) RangeWithCopy(f func(key, value string
 	}
 }
 
+func (h *requestOrResponseTrailerMapImpl) GetAllHeaders() map[string][]string {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	h.initTrailers()
+	copied_headers := make(map[string][]string)
+	for key, values := range h.headers {
+		copied_headers[key] = values
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/go/pkg/http/type.go:338]: this won't clone a slice too.

Comment (willemveerman) [reply contrib/golang/filters/http/source/go/pkg/http/type.go:338]: I don't fully understand, can you elaborate?

Comment (doujiang24) [reply contrib/golang/filters/http/source/go/pkg/http/type.go:338]: it is shallow copy now, here is an example:
```
	// slice
	s := []int{1, 2, 3}
	d := s
	s[0] = 4
	println(d[0]) // 4
```
I think we need deep clone for safety, thoughts?

Comment (willemveerman) [reply contrib/golang/filters/http/source/go/pkg/http/type.go:338]: I'm not using this approach any more `d := s`

So I don't think your example applies.

````
map_one := map[string][]string{"one": {"a", "b", "c"}, "two": {"twoa", "twob", "twoc"}}

fmt.Printf("map_one: %v\n", map_one) \\ map_one: map[one:[a b c] two:[twoa twob twoc]]

map_copy := make(map[string][]string)

for key, value := range map_one {
    map_copy[key] = value
}

map_one["one"] = []string{"a1"}

fmt.Printf("map_copy: %v\n", map_copy) \\  map_copy: map[one:[a b c] two:[twoa twob twoc]]

fmt.Printf("map_copy key 'one': %v\n", map_copy["one"]) \\  map_copy key 'one': [a b c]

fmt.Printf("map_one key 'one': %v\n", map_one["one"]) \\  map_one key 'one': [a1]

```

Comment (willemveerman) [reply contrib/golang/filters/http/source/go/pkg/http/type.go:338]: I see what you mean.... This should do what we need

```
	for key, value := range h.headers {
		copied_headers[key] = append(value[:0:0], value...)
	}
```

Or we could do this
```
	for key, value := range h.headers {
		copied_headers[key] = make([]string, len(value))
		copy(copied_headers[key], value)
	}

Comment (willemveerman) [reply contrib/golang/filters/http/source/go/pkg/http/type.go:338]: The second option is more readable imo

**Comment by doujiang24** [comment]
@willemveerman Here is a doc section: [Testing Envoy with Bazel](https://github.com/envoyproxy/envoy/tree/main/bazel#testing-envoy-with-bazel)

For the current ci failure: 
```
compilepkg: missing strict dependencies:
	/b/f/w/contrib/golang/filters/http/test/test_data/basic/filter.go: import of "golang.org/x/exp/slices"
```
https://dev.azure.com/cncf/envoy/_build/results?buildId=170352&view=logs&j=8c169225-0ae8-53bd-947f-07cb81846cb5&t=d1a98671-b7ba-5fbf-f06c-ff337c010df4&l=159

As the error message shows it missing dependency.
it would be a easier way to remove such dependency, it might be a bit complicated to introduce a golang dependency in bazel.

**Comment by willemveerman** [comment]
@doujiang24 OK, it's ready for review now

**Comment by willemveerman** [comment]
OK I've added one further test to cover the case of the headerMap being changed, in order to verify that the slices are being copied. Sorry for delay, I had a hard time building the tests on my m2 arm, but found a workaround. @doujiang24

File: contrib/golang/filters/http/source/go/pkg/http/type.go [hunk 6]
```diff
@@ -155,6 +155,18 @@ func (h *requestOrResponseHeaderMapImpl) RangeWithCopy(f func(key, value string)
 	}
 }
 
+func (h *requestOrResponseHeaderMapImpl) GetAllHeaders() map[string][]string {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	h.initHeaders()
+	copied_headers := make(map[string][]string)
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/go/pkg/http/type.go:162]: style: `camelCase` is better golang style, `copiedHeaders` would be better.

**Review by doujiang24 — COMMENTED** [review]
just a small style issue, otherwise, lgtm, thanks~

**Review by phlax — APPROVED** [review]

**Pull request merged**